src/query/expression/src/types/decimal.rs

```diff
@@ -63,10 +65,183 @@ pub struct DecimalSize {
+    }
+
+    fn to_column_from_buffer(value: Buffer<Self>, size: DecimalSize) -> DecimalColumn {
+        DecimalColumn::Decimal256(value, size)
+    }
+
+    fn try_downcast_column(column: &Column) -> Option<(Buffer<Self>, DecimalSize)> {
+        let column = column.as_decimal()?;
+        match column {
+            DecimalColumn::Decimal128(_, _) => None,
+            DecimalColumn::Decimal256(c, size) => Some((c.clone(), *size)),
+        }
+    }
+}
+
+static MAX_DECIMAL128_PRECISION: u8 = 38;
+static MAX_DECIMAL256_PRECISION: u8 = 76;
+
 impl DecimalDataType {
-    pub fn from_size(_size: DecimalSize) -> Result<DecimalDataType> {
-        // todo!("decimal")
-        Err(ErrorCode::Unimplemented("decimal"))
+    pub fn from_size(size: DecimalSize) -> Result<DecimalDataType> {
+        if size.precision < 1 || size.precision > MAX_DECIMAL256_PRECISION {
+            return Err(ErrorCode::Overflow(format!(
+                "Decimal precision must be between 1 and {}",
+                MAX_DECIMAL256_PRECISION
+            )));
+        }
+
+        if size.scale > size.precision {
+            return Err(ErrorCode::Overflow(format!(
+                "Decimal scale must be between 0 and precision {}",
+                size.precision
+            )));
+        }
+        if size.precision <= MAX_DECIMAL128_PRECISION {
+            Ok(DecimalDataType::Decimal128(size))
+        } else {
+            Ok(DecimalDataType::Decimal256(size))
+        }
+    }
+
+    pub fn scale(&self) -> u8 {
+        crate::with_decimal_type!(|DECIMAL_TYPE| match self {
+            DecimalDataType::DECIMAL_TYPE(size) => size.scale,
+        })
+    }
+
+    pub fn precision(&self) -> u8 {
+        crate::with_decimal_type!(|DECIMAL_TYPE| match self {
+            DecimalDataType::DECIMAL_TYPE(size) => size.precision,
+        })
+    }
+
+    pub fn max_precision(&self) -> u8 {
+        match self {
+            DecimalDataType::Decimal128(_) => MAX_DECIMAL128_PRECISION,
+            DecimalDataType::Decimal256(_) => MAX_DECIMAL256_PRECISION,
+        }
+    }
+
+    pub fn max_result_precision(&self, other: &Self) -> u8 {
+        if matches!(other, DecimalDataType::Decimal128(_)) {
+            return self.max_precision();
+        }
+        other.max_precision()
+    }
+
+    pub fn binary_result_type(
+        a: &Self,
+        b: &Self,
+        is_multiply: bool,
+        is_divide: bool,
+        is_plus_minus: bool,
+    ) -> Result<Self> {
+        let mut scale = a.scale().max(b.scale());
+        let mut precision = a.max_result_precision(b);
+
+        let multiply_precision = a.precision() + b.precision();
+        let divide_precision = a.precision() + b.scale();
+
+        // for addition/subtraction, we add 1 to the width to ensure we don't overflow
+        let plus_min_precision =
+            (a.precision() - a.scale()).max(b.precision() - b.scale()) + scale + 1;
+
+        if is_multiply {
+            scale = a.scale() + b.scale();
```

Comment:
need to check max scale?

Reply:
We will check it in `from_size` function.